runtime-modules/storage/src/lib.rs

```diff
@@ -3942,4 +3602,342 @@ impl<T: Trait> Module<T> {
+                        ))
+                        .with_distribution_buckets(
+                            DistributionBucketPicker::<T>::pick_distribution_buckets(
+                                dyn_bag_id.into(),
+                            ),
+                        ))
+                },
+                |_| Err(Error::<T>::DynamicBagExists.into()),
+            )
+        } else {
+            Self::ensure_bag_exists(&bag_op.bag_id)
+        }
+    }
+
+    /// Utility function that does the bulk of bag / data objects operation
+    fn try_mutating_storage_state(
+        account_id: T::AccountId,
+        bag_op: BagOperation<T>,
+    ) -> DispatchResult {
+        // attempt retrieve existing or create new one if operation is create
+        let Bag::<T> {
+            stored_by,
+            distributed_by,
+            deletion_prize,
+            objects_total_size,
+            objects_number,
+        } = Self::retrieve_dynamic_bag(&bag_op)?;
+
+        // check and generate any objects to add/remove from storage
+        let object_creation_list = Self::construct_objects_to_upload(&bag_op)?;
+        let objects_removal_list = Self::construct_objects_to_remove(&bag_op)?;
+
+        // new candidate voucher
+        let new_voucher_update = VoucherUpdate {
+            objects_total_size,
+            objects_number,
+        }
+        .add_objects_list(object_creation_list.as_slice())
+        .sub_objects_list(objects_removal_list.as_slice());
+
+        // new candidate storage buckets
+        let new_storage_buckets =
+            Self::generate_new_storage_buckets(&stored_by, new_voucher_update, &bag_op.params)?;
+        // new candidate distribution buckets
+        let new_distribution_buckets =
+            Self::generate_new_distribution_buckets(&distributed_by, &bag_op.params);
+
+        // check that user or treasury account have enough balance
+        let (net_prize, storage_fee) = Self::ensure_sufficient_balance(
+            &bag_op,
+            new_voucher_update,
+            deletion_prize,
+            &account_id,
+        )?;
+
+        //
+        // == MUTATION SAFE ==
+        //
+
+        // insert candidate storage buckets: no op if new_storage_buckets is empty
+        new_storage_buckets.iter().for_each(|(id, bucket)| {
+            StorageBucketById::<T>::insert(&id, bucket.clone());
+            Self::deposit_event(RawEvent::VoucherChanged(*id, bucket.voucher.clone()));
+        });
+
+        // insert candidate distribution buckets: no op if new_distribution_buckets is empty
+        new_distribution_buckets.iter().for_each(|(id, bucket)| {
+            DistributionBucketByFamilyIdById::<T>::insert(
+                &id.distribution_bucket_family_id,
+                &id.distribution_bucket_index,
+                bucket,
+            )
+        });
+
+        // FIRST refund or request deletion prize first: no-op if amnt = 0
+        match net_prize {
```

Comment:
These calls should be made first after the mutation safe mark. Because they are potentially fallible.

Reply:
Even when I already checked that Treasury/signer account have enough balance?

Reply:
:+1:

Reply:
This code has no dependency on the previous code and we can move it freely. However, if the check you mentioned would be changed or removed (by mistake) the first position of the snippet would make the method still safe in contrast with the current situation.